Isomorphic timeline events viz 34924

markbastian opened the pull request:
This primarily required adding logic as found in `/api/timeline?include=events` to `s/defmethod render :isomorphic :- common/RenderedPulseCard`. This is done by adding one extra accretion function, `add-dashcard-timeline-events` to the logic and making code that expected a dashboard also take a dashcard.

Resolves #34924

markbastian committed:
Initial states of isomorphic line charts/events for BE

Initial stages of getting line charts, with a focus towards timeline events,
rendered using the isomorphic pipeline. This change handles cases where a
card has multiple breakouts, forming multiple timelines, vs. a dashcard, which
may have both a query and series to establish multiple timelines.

adam-james-v commented on dev/src/dev/render_png.clj:
```diff
@@ -2,16 +2,17 @@
+    [metabase.pulse.render.test-util :as render.tu]
+    [metabase.query-processor :as qp]
+    [metabase.query-processor.middleware.permissions :as qp.perms]
+    [toucan2.core :as t2]))
```
Good to actually have the proper require in here. Not sure how I missed that before.

adam-james-v committed:
Timeline events passed to :isomorphic chart type

Timeline events are now grabbed for the card(s) and passed into the static-viz rendering pipeline.

This commit adds:

 - `dashcard-timeline-events` helper fn that gets events for every timeline in the Collection that the card is part
 of.
 - adjustments to the `execute-multi-card` function for the situation where it's called with `nil`
 dashcard-or-id.
   - This function may need to be re-visited as the isomorphic chart work continues (namely, we need to make sure
 this function works correctly for all scenarios 1. single card 2. single card with multiple series (a 'combo' chart)
 3. dashcard with multiple single series visualized on one card and 4. dashcard with mixed single/multi-series
 cards (not sure if this last one is possible).
 - a test to ensure timeline events are added

adam-james-v commented on src/metabase/pulse/util.clj:
```diff
@@ -14,6 +14,7 @@
    [toucan2.core :as t2]))
 
 ;; TODO - this should be done async
+;; TODO - this and `execute-multi-card` should be made more efficient: eg. we query for the card several times
```
I think that in order to make these execute fns more efficient we should change the contract to require passing in the instance, so that we fetch earlier and only once.

Not needed for this PR, but an idea anyway.

adam-james-v reviewed (approved):
These changes look good. Mark and I paired on some of this as well, so I'm familiar with the code added already.
